src/Integration.Vsix/Delegates/SonarAnalyzerManager.cs

```diff
@@ -0,0 +1,135 @@
+using SonarLint.VisualStudio.Integration.Suppression;
+using SonarLint.VisualStudio.Integration.Vsix.Suppression;
+using SonarQube.Client.Services;
+
+namespace SonarLint.VisualStudio.Integration.Vsix
+{
+    // This class is simply responsible to react to the solution binding state change and to trigger the right workflow
+    internal sealed class SonarAnalyzerManager : IDisposable
+    {
+        private readonly IActiveSolutionBoundTracker activeSolutionBoundTracker;
+        private readonly ISonarQubeService sonarQubeService;
+        private readonly Workspace workspace;
+        private readonly IQualityProfileProvider qualityProfileProvider;
+        private readonly IVsSolution vsSolution;
+        private readonly ILogger logger;
+
+        internal /* for testing purposes */ SonarAnalyzerWorkflowBase currentWorklow;
+        private SonarQubeIssuesProvider sonarQubeIssueProvider;
+
+        public SonarAnalyzerManager(IActiveSolutionBoundTracker activeSolutionBoundTracker, ISonarQubeService sonarQubeService,
+            Workspace workspace, IQualityProfileProvider qualityProfileProvider, IVsSolution vsSolution, ILogger logger)
+        {
+            if (activeSolutionBoundTracker == null)
+            {
+                throw new ArgumentNullException(nameof(activeSolutionBoundTracker));
+            }
+            if (sonarQubeService == null)
+            {
+                throw new ArgumentNullException(nameof(sonarQubeService));
+            }
+            if (workspace == null)
+            {
+                throw new ArgumentNullException(nameof(workspace));
+            }
+            if (qualityProfileProvider == null)
+            {
+                throw new ArgumentNullException(nameof(qualityProfileProvider));
+            }
+            if (vsSolution == null)
+            {
+                throw new ArgumentNullException(nameof(vsSolution));
+            }
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            this.activeSolutionBoundTracker = activeSolutionBoundTracker;
+            this.sonarQubeService = sonarQubeService;
+            this.workspace = workspace;
+            this.qualityProfileProvider = qualityProfileProvider;
+            this.vsSolution = vsSolution;
+            this.logger = logger;
+
+            this.activeSolutionBoundTracker = activeSolutionBoundTracker;
+            activeSolutionBoundTracker.SolutionBindingChanged += OnSolutionBindingChanged;
+
+            RefreshWorkflow(this.activeSolutionBoundTracker.CurrentConfiguration);
+        }
+
+        private void OnSolutionBindingChanged(object sender, ActiveSolutionBindingEventArgs e)
+        {
+            RefreshWorkflow(e.Configuration);
+        }
+
+        private void RefreshWorkflow(BindingConfiguration configuration)
+        {
+            this.sonarQubeIssueProvider?.Dispose();
+            this.currentWorklow?.Dispose();
```

Comment:
Probably worth adding a comment to say there is a _potential_ race condition here, if an analysis is in progress while we are clearing and re-setting the delegates. It it does occur then the results for the in-progress could be wrong, but that any inconsistencies should be fixed on the next analysis run (we'll forget otherwise, and a comment would at least give us a hint in case we do have to diagnose weird results).